src/coreclr/jit/lclmasks.cpp

```diff
@@ -0,0 +1,498 @@
+// LclMasksCheckVisitor: Find all lcl var definitions and uses. For each one, update the weighting.
+//
+class LclMasksCheckVisitor final : public GenTreeVisitor<LclMasksCheckVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LclMasksCheckVisitor(Compiler* compiler, weight_t bbWeight, LclMasksWeightTable* weightsTable)
+        : GenTreeVisitor<LclMasksCheckVisitor>(compiler)
+        , bbWeight(bbWeight)
+        , weightsTable(weightsTable)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTreeHWIntrinsic* convertOp = nullptr;
+
+        bool isLocalStore  = false;
+        bool isLocalUse    = false;
+        bool isInvalid     = false;
+        bool hasConversion = false;
+
+        switch ((*use)->OperGet())
+        {
+            case GT_STORE_LCL_VAR:
+                isLocalStore = true;
+
+                // Look for:
+                //      use:STORE_LCL_VAR(ConvertMaskToVector(x))
+
+                if ((*use)->AsLclVar()->Data()->OperIsConvertMaskToVector())
+                {
+                    convertOp     = (*use)->AsLclVar()->Data()->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            case GT_LCL_VAR:
+                isLocalUse = true;
+
+                // Look for:
+                //      user:ConvertVectorToMask(use:LCL_VAR(x)))
+
+                if (user->OperIsConvertVectorToMask())
+                {
+                    convertOp     = user->AsHWIntrinsic();
+                    hasConversion = true;
+                }
+                break;
+
+            case GT_LCL_ADDR:
+                isInvalid = true;
+                break;
+
+            default:
+                break;
+        }
+
+        if (isLocalStore || isLocalUse || isInvalid)
+        {
+            GenTreeLclVarCommon* lclOp = (*use)->AsLclVarCommon();
+
+            // Get the existing weighting (if any).
+            LclMasksWeight  defaultWeight;
+            LclMasksWeight* weight = weightsTable->LookupPointerOrAdd(lclOp->GetLclNum(), defaultWeight);
+
+            // Update the weights.
+            JITDUMP("Local %s V%02d at [%06u] ", isLocalStore ? "store" : "var", lclOp->GetLclNum(),
+                    m_compiler->dspTreeID(lclOp));
+            if (isInvalid)
+            {
+                JITDUMP("cannot be converted. ");
+                weight->InvalidateWeight();
+            }
+            else
+            {
+                JITDUMP("has %s conversion. ", hasConversion ? "mask" : "no");
+                weight->UpdateWeight(isLocalStore, hasConversion, bbWeight);
+            }
+
+            // Cache the simd type data of the conversion.
+            if (hasConversion)
+            {
+                assert(convertOp != nullptr);
+                weight->CacheSimdTypes(convertOp);
```

Comment:
Looks like one of my comments got lost somewhere, but I don't think this can just overwrite the type -- it's possible to have a mismatch here using `Unsafe.BitCast`, in which case I think this needs to indicate that somehow and skip the conversion in the second pass.

Reply:
There's a couple of comments above I still need to do - just been distracted with other things :)